File: server-python/src/routes/channels.py
from flask import Blueprint, request, jsonify
from services.youtube_service import YouTubeService
from utils.logger import logger
from middleware.auth import authenticate, require_permission
from middleware.rbac import require_scopes

bp = Blueprint('channels', __name__, url_prefix='/api/channels')
youtube_service = YouTubeService()

@bp.route('/search', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def search_channels():
    query = request.args.get('query')
    if not query:
        return jsonify({'error': 'Query parameter is required'}), 400
    
    logger.info(f'Searching channels for query: {query}')
    channels = youtube_service.search_channels(query)
    return jsonify(channels)

@bp.route('/<string:channel_id>', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def get_channel_info(channel_id):
    logger.info(f'Getting info for channel: {channel_id}')
    channel_info = youtube_service.get_channel_info(channel_id)
    if not channel_info:
        return jsonify({'error': 'Channel not found'}), 404
    return jsonify(channel_info)

@bp.route('/<string:channel_id>/videos', methods=['GET'])
@authenticate
@require_permission('analysis:read')
@require_scopes(['read'])
def get_channel_videos(channel_id):
    logger.info(f'Getting videos for channel: {channel_id}')
    videos = youtube_service.get_channel_videos(channel_id)
    return jsonify(videos)

File: server/src/routes/channels.js
const express = require('express');
const youtubeService = require('../services/youtubeService');
const { authenticate, requireScopes } = require('../middleware/auth');
const { requirePermission } = require('../middleware/rbac');
const logger = require('../utils/logger');

const router = express.Router();

// Search for channels
router.get('/search',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res, next) => {
    try {
      const { q: query, maxResults = 25, minSubs = 10000, maxSubs = 500000 } = req.query;

      if (!query) {
        return res.status(400).json({
          success: false,
          message: 'Query parameter is required',
        });
      }

      const channels = await youtubeService.searchChannels(query, parseInt(maxResults), {
        min: parseInt(minSubs),
        max: parseInt(maxSubs),
      });

      res.json({
        success: true,
        channels,
        count: channels.length,
      });

    } catch (error) {
      next(error);
    }
  });

// Get channel information
router.get('/:channelId',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res, next) => {
    try {
      const { channelId } = req.params;
      const channelInfo = await youtubeService.getChannelInfo(channelId);

      res.json({
        success: true,
        channel: channelInfo,
      });

    } catch (error) {
      next(error);
    }
  });

// Get channel videos
router.get('/:channelId/videos',
  authenticate,
  requirePermission('analysis:read'),
  requireScopes(['read']), // For API key access
  async (req, res, next) => {
    try {
      const { channelId } = req.params;
      const { maxResults = 10, timeWindow = 7 } = req.query;

      const publishedAfter = new Date();
      publishedAfter.setDate(publishedAfter.getDate() - parseInt(timeWindow));

      const videos = await youtubeService.getChannelVideos(
        channelId,
        parseInt(maxResults),
        publishedAfter.toISOString(),
      );

      res.json({
        success: true,
        videos,
        count: videos.length,
      });

    } catch (error) {
      next(error);
    }
  });

module.exports = router;
